[Poker game step5] : 포커게임 결과 화면

1Consumption opened the pull request:
* 고민한 점

  1. getter를 사용하여 내부 변수에 접근하지 말고 동작을 전달하여 원하는 결과를 얻기
     * private한 변수에 접근하려고 클래스 내에서 getter를 작성하였습니다. 하지만 내부 프로퍼티를 외부 객체에서 접근하지말게 하라는 마스터의 피드백이 있었고, 이를 고치려고 노력했습니다. 마스터의 힌트는 동작을 전달하라는 것이었는데, 이 부분에서 클로저를 사용 해야겠다는 생각을 했습니다. 클래스마다 forEach 펑션을 만들어주고 동작을 넘겨 받게끔 하였습니다. 하지만 이 경우 클래스별로 중복되는 코드가 발생하기에 'getter가 더 효율적이지않나' 라는 생각을 가졌었으나, 클래스에서 동작을 정의하지않고 동작을 넘겨받은 것을 실행하기 때문에 재사용성 면에서 크게 뛰어난 것을 알게 되었습니다.
     * 또한 이 과정에서 추상화에 대해 생각해보게 되었습니다. 예를들어 플레이어는 카드가 몇장이 있는지보다 카드가 있다는 것만 알면 되기 때문에 기존에 Player 클래스에서 Card 객체 배열을 가지고 있었으나, 이를 Card 배열을 변수로 가지는 Hand 클래스를 만들고, Player에 포함시켰습니다.
  2. stackView 내부 뷰의 constraint 설정
     * stackView에 view들을 넣고 내부 view에 constraint 값을 줘서 내부 view를 배치하려고 하였습니다. 하지만 뜻대로 되지 않았습니다. stackView는 동일한 규칙으로 view를 배치하는 형태이기 때문에 내부 view에 constraint를 줘도 적용되지 않는다는 점을 알게 되었고, 이를 통해 stackView에 포함시킬 view와 포함시키지 않을 view를 분류하는 기준을 알게 되었습니다.

* 아쉬운 점

  * 클래스간 의존도를 낮추려고 노력하였으나, 아직 많이 부족하다고 생각됩니다.

  * switch문이 너무 많다고 생각됩니다.

  * ViewController 클래스의 라인이 너무 많습니다.

1Consumption committed:
Card 이미지 추가

1Consumption committed:
Dealer가 Player를 상속하게 코드 수정

기존의 코드는 프로토콜을 채용하였으나, 프로퍼티가 공통되는 경우는 상속이 조금 더 적합하기 때문에 상속으로 코드 수정

1Consumption committed:
PokerGame 클래스 불필요 코드 제거 및 수정

1Consumption committed:
Hand 클래스 생성

Player의 card를 관리하는 클레스

1Consumption committed:
Hand, Player 클래스 수정

내부에 forEach 메소드를 생성하여 동작을 넘겨받고, 내부에서 실행하도록 함

1Consumption committed:
Players 클래스 추가

PokerGame 클래스 내의 player 배열 변수인 players를 대체할 예정

1Consumption committed:
Players, PokerGame 클래스 변경

Players에 forEach 메소드를 추가하여 동작을 넘겨받아 행동하도록 기능 구현

1Consumption committed:
PokerGame 클래스 기능 추가

forEachDealer, forEachPlayers 메소드를 추가하여 private한 프로퍼티에 동작을 전달하여 최종적으로 viewController 클래스에서 사용 가능하도록 수정

1Consumption committed:
불필요한 메소드 제거

1Consumption committed:
카드 이름을 받아서 imageView 만들도록 수정

PokerGame 내 createCard() 메소드에 String 타입의 cardName 매개변수를 넘겨줘서 cardName에 해당하는 카드 이미지 생성하게끔 함.
Card 클래스의 Suit enum의 rawValue를 각 suit의 첫번째 알파벳으로 변경하여 cardName과 매칭할수 있게됨.

1Consumption committed:
ViewController에 UISegmentedControl 인스턴스 추가

Stud와 플레이어 수를 선택할 수 있게 해주는 view 역할을 함

1Consumption committed:
stackView, segmentedControl 생성 함수 추가

1Consumption committed:
createView 메소드 생성

createStackView, createSegmentedControl 메소드를 한번에 호출하여 UIView 인스턴스를 할당해줌

1Consumption committed:
segmentedControl 추가

stud과 플레이어 수를 결정할 view 역할인 segmentControl을 추가하였음. 기존에 스택뷰를 사용하였으나 스택뷰의 역할과 동떨어져 바로 superView에 add 하였음

1Consumption committed:
UIView extension 추가 및 함수 구조 변경

UIView 객체의 constraint를 조정하기 쉽도록 extension 추가 함

addViewsIntoSuperView() 메소드에서 만들었던 view를 superView에 추가해줌
setupViews() 메소드에서 view들의 constraint를 설정해 줌.

1Consumption committed:
segmentedControl에 기능 추가

실제 선택 값에 따라 PokerGame의 stud, numOfPlayer 값을 바꿔줌

1Consumption committed:
card들 화면에 띄워줌

기본 stud 7, 플레이어 수 4를 기준으로 카드를 생성해서 view에 표시해줌

1Consumption committed:
segmentedControl과 card 연결

segmentedControl 을 조작하여 값을 바꾸면 바꾼 값이 cardView에 적용됨

1Consumption committed:
player인지 dealer인지 표시

playerStackView에 Label을 추가해서 player인지 딜러인지 표시해줌.

1Consumption committed:
shake gesture 추가

디바이스를 흔들면 카드를 섞고 다시 view에 보여줌

1Consumption committed:
승자 표시 이미지 추가

1Consumption committed:
readme 수정

추가한 기능 정리
스크린샷 첨부

godrm commented on CardGameApp/CardGameApp/ViewController.swift:
```diff
@@ -9,47 +9,176 @@
+        default:
+            stud = .sevenCardStud
+            break
+        }
```
여기 switch-case 구문도 stud 타입의 init 으로 대체해보세요.

godrm commented on CardGameApp/CardGameApp/ViewController.swift:
```diff
@@ -9,47 +9,176 @@
+                addCardIntoStackView(stackView: cardStackView, cardName: "\($0)")
+            }
+            playerStackView.addArrangedSubview(cardStackView)
+            pokerStackView.addArrangedSubview(playerStackView)
```
클로저를 활용해서 잘 구현했네요.
클로저 내부에 모든 동작을 다 적어야만 하는 건 아닙니다. 동작에 대한 메소드를 만들고 호출해도 되죠. 좀 길어진다 싶으면 단순화하는 게 필요하다는 의미입니다. 
매개 변수를 활용하지 않고 `$0`를 사용했는데, player 라는 걸 쉽게 인지하지 못할 수도 있다는 단점이 있습니다.  
특히 이 부분은 $0 가 중첩되서 더 오해의 소지가 있습니다. 참고하세요.
```
$0.forEach{
                 addCardIntoStackView(stackView: cardStackView, cardName: "\($0)")
}
```

godrm replied:
조금 길지만 그만큼 구체적이네요. 뒤에 PlayerHas는 빼도 되지 않을까 싶네요. 참고하세요.

1Consumption replied:
넵 감사합니다!

godrm commented on CardGameApp/CardGameApp/model/Dealer.swift:
```diff
@@ -8,33 +8,16 @@
 
-class Dealer{
+class Dealer:Player{
     private var cardDeck:CardDeck
```
어떤 경우는 선언에 `:` 앞 뒤에 공백을 붙이기도 하고 안 붙이기도 하네요.
일관성 있게 표현하는 게 좋습니다. 그래야 코드를 읽을 때 흐름이 끊기거나 어색하지 않습니다.

godrm commented on CardGameApp/CardGameApp/model/Hand.swift:
```diff
@@ -0,0 +1,27 @@
+import Foundation
+
+class Hand{
+    private var card:[Card] = [Card]()
```
CardDeck 과 Hand 모두 [Card]를 내부에서 처리하는 클래스인데 어떻게 다른건가요?
공통적인 부분도 많지 않나요?

1Consumption replied:
CardDeck은 전체 카드를 가지고 있습니다. 카드를 분배할 뿐 받지는 않습니다. 반면 Hand는 플레이어의 카드를 가지고 있으며 카드를 받습니다. 물론 어느정도 중복되는 기능은 있다고 생각됩니다만, CardDeck은 struct여서 상속관계를 가질 수 없고, 프로토콜을 적용하기에는 [Card] 프로퍼티를 가지기에 적합하지 않다고 생각되어 CardDeck과 Hand를 따로 분리하였습니다.

godrm commented on CardGameApp/CardGameApp/model/Player.swift:
```diff
@@ -9,26 +9,20 @@
+    func forEach(_ transform : (Card) -> ()){
+        hand.forEach{
+            transform($0)
+        }
```
forEach는 예시로 들었던 이름인데 꼭 이걸 따라야 하는 건 아닙니다. 
Player 에게 forEach는 어떤 걸 의미하는 걸까요? 좀 더 명확하게 표현해보세요.

godrm commented on CardGameApp/CardGameApp/model/PokerGame.swift:
```diff
@@ -36,29 +36,29 @@ class PokerGame{
+        readyPlayer()
+        stud.forEach {
+            players.forEach{
+                dealCardToPlayer(player: $0)
```
클로저를 사용해도 이렇게 깊이가 깊어지면 분리하는 게 좋습니다. 
함수가 클로저를 대신할 수도 있습니다. 참고하세요.

godrm commented on CardGameApp/CardGameApp/model/PokerGame.swift:
```diff
@@ -36,29 +36,29 @@ class PokerGame{
     }
     
-    private func dealCardToPlayer(player:MustHaveCard){
+    private func dealCardToPlayer(player:Player){
```
이런 함수 경우에는 `dealCard(to player: Player)` 형태로 선언해서 중복을 줄여주는 게 읽기 편합니다.

godrm commented on CardGameApp/CardGameApp/model/PokerGame.swift:
```diff
@@ -36,29 +36,29 @@ class PokerGame{
+    private func dealCardToPlayer(player:Player){
         let myCard = dealer.deal()
         if let card = myCard{
             player.receiveCard(card: card)
```
아래 else 구문은 의미가 없습니다.
앱에서는 print()를 사용하지 마세요. 
카드가 없으면 사용자에게 알려줘 흐름을 제어할 수 있어야 하는데 print는 개발자에게만 의미가 있습니다.

godrm commented on CardGameApp/CardGameApp/model/PokerGame.swift:
```diff
@@ -70,26 +70,12 @@ class PokerGame{
         self.stud = stud
         self.numOfPlayer = numOfPlayer
-        readyPlayer()
         gameStart()
```
힘수 이름을 setGameStyle() 이라서 스타일을 설정만 하는 것 같은데 내부 동작은 gameStart()도 포함하고 있네요. 
함수 이름이 다른 동작을 표현하는 걸까요? 동작을 여기서 하지 않는게 좋은 걸까요?

1Consumption replied:
stud나 플레이어 수를 변경하면 그에 맞게 플레이어를 생성하고, 카드를 분배해야 하기 때문에 플레이어를 생성하고 카드를 분배하는 역할을 하는 gameStart 메소드를 포함했습니다. 하지만 이 경우 하나의 메소드에서 두가지의 동작을 하기에 분리하는 것이 좋아보입니다. 수정하겠습니다.

godrm commented on CardGameApp/CardGameApp/model/card.swift:
```diff
@@ -17,10 +17,10 @@ class Card{
+        case spade = "s"
+        case clover = "c"
+        case heart = "h"
+        case diamond = "d"
```
왜 유니코드에서 ascii로 바꾼건가요?

1Consumption replied:
카드와 이미지 이름을 매칭해주기 위해 Card 클래스에 CustomStringConvertible 프로토콜을 적용하여 문자열로 표현하였습니다. 유니코드로 rawValue를 주면 결국 다시 String 타입으로 캐스팅 해야하기 때문에 String 타입으로 선언하였습니다.

godrm replied:
이런 작업을 할 때 중요한 포인트는 Card 처럼 모델 객체가 어떤 타입으로 값을 보관하느냐입니다. 
유니코드도 String 인데 타입이 달라진 건 아닌 것 같고, 말한데로 이미지 이름과 매칭해주려고 그런 것 같네요. 
데이터를 표현할 때 가장 적합한 데이터 구조와 뷰에 출력하기 위한 데이터 구조가 항상 일치하지 않을수도 있습니다. 
왜냐하면 데이터를 처리하는 로직이 중요하고, 거기에 맞는 타입을 갖는 거고
뷰에 출력할 때는 어떤 뷰에 어떻게 보여주느냐 요구사항에 따라 바뀔 수도 있는 부분입니다. 
뷰 형태가 바뀔때마다 데이터 구조를 바뀔 수는 없겠죠?

1Consumption replied:
그렇다면 이 경우는 유니코드에 맞게 값을 매핑해주는 것이 좋을까요?

godrm replied:
지금은 상관없습니다. 타입을 바꾸려고 생각한 과정에 대한 피드백이었습니다. 별차이가 없다는 거죠.

1Consumption replied:
넵 확인했습니다. 앞으로 데이터 구조에 대해 한번 더 생각해보겠습니다.

godrm reviewed (commented):
구조적으로 나아진 부분이 많지만, 세세하게 개선할 부분을 피드백으로 남겼습니다.

1Consumption committed:
코드 스타일 수정

1Consumption committed:
함수 이름 중복 제거

godrm commented:
수정을 모두 다 한 상태인가요?

1Consumption commented:
아니요, 아직 하는 중입니다!

1Consumption committed:
setupPokerStackView() 메소드 단순화

-zsh로 표시하던 Player 오브젝트를 player 변수에 담아 알아보기 쉽게 표현
label을 생성하던 코드를 메소드로 따로 빼서 동작 단순화
Player객체에 order 변수와 CustomStringConvertible 프로토콜을 추가하여 label을 생성하는 메소드에서 타입을 비교하지않고 바로 Player 인스턴스를 문자열로 label에 출력

1Consumption committed:
불분명한 forEach 메소드 이름을 구체화

1Consumption committed:
setGameStyle 메소드에서 gameStart 메소드 제거

하나의 함수에서 하나의 일만 하도록 수정

1Consumption committed:
SegmentedConrol handler 내 switch문 제거

stud, numOfPlayer enum에 CaseIterable 프로토콜을 채용하고 init 메소드를 만들었음. 이 init 메소드에서는 SegmentedControl에서 선택한 아이템의 인덱스를 파라미터로 받고, Stud, NumOfPlayer를 배열화 한 값에 파라미터를 대입하여 할당해줌

1Consumption committed:
불필요한 조건문 제거

카드를 더 뽑는 경우가 없기 때문에 카드가 없는 경우는 고려하지 않음

1Consumption commented:
피드백 반영하여 수정했습니다!
